modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,238 @@
+  enqueuedAtOrdinal: SnapshotOrdinal,
+  sendsSoFar: NonNegLong
+)
+
+case class State(
+  pending: Queue[TrackedBinary],
+  cap: NonNegLong,
+  retryMode: Boolean,
+  noConfirmationsSinceRetryCount: NonNegLong,
+  backoffExponent: NonNegLong
+)
+
+object State {
+  def empty: State = State(
+    pending = Queue.empty[TrackedBinary],
+    cap = 4L,
+    retryMode = false,
+    noConfirmationsSinceRetryCount = 0L,
+    backoffExponent = 0L
+  )
+}
+
+trait StateChannelBinarySender[F[_]] {
+  def processPending: F[Unit]
+  def clearPending: F[Unit]
+  def confirm(globalSnapshot: GlobalIncrementalSnapshot): F[Unit]
+  def process(binaryHashed: Hashed[StateChannelSnapshotBinary]): F[Unit]
+}
+
+object StateChannelBinarySender {
+  def make[F[_]: Async: Hasher](
+    identifierStorage: IdentifierStorage[F],
+    globalL0ClusterStorage: L0ClusterStorage[F],
+    lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+    stateChannelSnapshotClient: StateChannelSnapshotClient[F]
+  ): F[StateChannelBinarySender[F]] =
+    Ref
+      .of[F, State](State.empty)
+      .map(make[F](_, identifierStorage, globalL0ClusterStorage, lastGlobalSnapshotStorage, stateChannelSnapshotClient))
+
+  def make[F[_]: Async: Hasher](
+    stateR: Ref[F, State],
+    identifierStorage: IdentifierStorage[F],
+    globalL0ClusterStorage: L0ClusterStorage[F],
+    lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+    stateChannelSnapshotClient: StateChannelSnapshotClient[F]
+  ): StateChannelBinarySender[F] =
+    new StateChannelBinarySender[F] {
+
+      private val noConfirmationsToTriggerRetryMode: PosLong = 5L
+      private val confirmedCountMultiplier: PosLong = 4L
+
+      private val logger = Slf4jLogger.getLogger
+
+      def process(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] =
+        lastGlobalSnapshotStorage.getOrdinal.map(_.getOrElse(SnapshotOrdinal.MinValue)).flatMap { currentOrdinal =>
+          stateR.modify {
+            case state @ State(pending, _, retryMode, _, _) =>
+              val updatedPending = pending :+ TrackedBinary(binary, enqueuedAtOrdinal = currentOrdinal, 0L)
+              val action =
+                if (retryMode) logger.warn(s"[RetryMode] Snapshot binary of hash ${binary.hash} enqueued.")
+                else
+                  post(binary).flatTap(_ => logger.info(s"Snapshot binary of hash ${binary.hash} enqueued and sent to GL0"))
+              val newState = state.copy(pending = updatedPending)
+              (newState, action)
+          }.flatMap(identity)
+        }
+
+      def confirm(globalSnapshot: GlobalIncrementalSnapshot): F[Unit] = for {
+        identifier <- identifierStorage.get
+        binaries = globalSnapshot.stateChannelSnapshots.get(identifier).toList.flatMap(_.toList)
+        confirmedHashesInGlobalSnapshot <- binaries.traverse(_.toHashed).map(_.map(_.hash))
+
```

Comment:
That is a very long lambda function for updating `stateR`. Can it be simplified/decomposed?

Reply:
@sdavidp What you mean by decomposing it? The fn itself is decomposed into smaller parts and variables. Would moving the internal fn to a separate function work for you?

```scala
_ <- stateR.update(updateState(dep1, dep2))

[...]

def updateState(dep1, dep2)(state): NewState = {
 ...
}
```

Reply:
It looks like the lambda is performing calculations for each field of the state object so maybe if that logic was decoupled it would read better? I'll leave it to you two to decide.

Reply:
We could either extract the pure part as @marcinwadon suggested or move each of the cases to a separate function but don't ask me because I may be biased as I wrote the statements and for me it is readable.